### ISSUE TITLE
backend for gitlab

### ISSUE THREAD

**Issue by divansantana** [arbox/org-sync#20]
Hi,

Would be really awesome if there was a backend for GitLab.

**Comment by arbox** [comment]
We'll try to provide one as soon as time required is there :)

**Comment by divansantana** [comment]
:+1:

**Comment by To1ne** [comment]
I was totally not aware of this project.

I've been working on a stand-alone package to synchronize gitlab issues/MRs to org-mode: https://gitlab.com/to1ne/org-gitlab

I will have a look at this package to see if would be interested in writing a GitLab backend for org-sync.

**Comment by arbox** [comment]
@To1ne you are welcome to add the Gitlab backend!

**Comment by To1ne** [comment]
@arbox Yeah, I'll be looking into org-sync some deeper. The plans I had for org-gitlab are a bit different, but maybe org-sync can also benefit from them. So bringing it all in might be a good call.

Are you still actively working on org-sync?

**Comment by arbox** [comment]
@To1ne only sporadically, time is too limited :(  If you want to help you are always welcome!

**Comment by yisraeldov** [comment]
Hello all, I have this almost working now.

It is missing the ability to update tasks, ( but it should be simple enough to implement )

It is just syncing the basic info.

https://github.com/yisraeldov/org-sync/tree/ydl-gitlab-sync

I hope to finish tomorrow to get the update working.  

( this is my first emacs package so, go easy )

**Comment by divansantana** [comment]
@yisraeldov good luck! I'll be keen to try it out. I'll try look at testing it sometime.

**Comment by yisraeldov** [comment]
@divansantana @To1ne 

This is now working, you can see the PR #32

**Comment by yisraeldov** [comment]
@arbox have you had a chance to look at the gitlab backend #32

**Comment by divansantana** [comment]
@yisraeldov so far with brief quick testing.

If one assigns a due date on an issue in gitlab, one gets this error: `2017-12-21`

Also I set `(setq org-sync-gitlab-domain "git.example.com")` but when running `org-sync-import` I got `No base-url backend available.`. I checked out the var `org-sync-backend-alist`. It was set to

```lisp
PERSONAL
org gitlab sync
amp stuff
my present and new laptop?
wifi
org agenda everything unscheduled
(("github.com/\\(?:repos/\\)?[^/]+/[^/]+" . org-sync-github-backend)
 ("bitbucket.org/[^/]+/[^/]+" . org-sync-bb-backend)
 ("gitlab.com" . org-sync-gitlab-backend)
 ("/projects/[^/]+" . org-sync-rmine-backend)
 ("rememberthemilk.com" . org-sync-rtm-backend))
```
So it still had `gitlab.com` set. Manually setting that var fixed the issue.

Other feedback.

I see labels and milestones don't pull through. Labels would obviously be very nice.

I successfully testing pulling of lots of issues!

I could also sync back.

I am however seeing a not so nice possible bug.

I see entries like this:

```
* Tasks
:PROPERTIES:
:url:      git.example.com/F3333000/puppet-main
:END:
** OPEN OPEN OPEN first test issue
:PROPERTIES:
:web-url:  "https://git.example.com/F3333000/puppet-main/issues/1"
:id:       1
:date-modification: 2017-12-07T19:01:33+0200
:date-creation: 2017-12-07T18:33:02+0200
:END:
: description first test issue
** OPEN OPEN OPEN second issue
:PROPERTIES:
:web-url:  "https://git.example.com/F3333000/puppet-main/issues/2"
:id:       2
:date-modification: 2017-12-07T19:01:33+0200
:date-creation: 2017-12-07T18:33:27+0200
:END:
: descr 2nd issue
** OPEN OPEN org created this issue
:PROPERTIES:
:web-url:  "https://git.example.com/F3333000/puppet-main/issues/3"
:id:       3
:date-modification: 2017-12-07T19:01:33+0200
:date-creation: 2017-12-07T18:58:29+0200
:END:
: Hello world. How are you `today`.
: 
: Bye.
** OPEN third issue created in gitlab
:PROPERTIES:
:web-url:  "https://git.example.com/F3333000/puppet-main/issues/4"
:id:       4
:date-modification: 2017-12-07T19:01:32+0200
:date-creation: 2017-12-07T19:01:28+0200
:END:
: hello
```
It seems to be duplicating the state of the issue, like `OPEN OPEN` and `OPEN OPEN OPEN`

**Comment by yisraeldov** [comment]
@divansantana 

Weird I don't get that bug. Try using the version from git-hub master. 
Labels would be nice, but I want to get the basic functionality merged in first.

What s the title when you go into your gitlab site ?

**Comment by divansantana** [comment]
So a couple of things.

Firstly this is all really awesome. Though still early days. 

Unfortunately in general it seems there's quite a bit of issues with org-sync in
general so hopefully the project picks up activity and gets some attention.

I found the gitlab backend pretty fast.

> If one assigns a due date on an issue in gitlab, one gets this error:
> 2017-12-21 and the syncing completely doesn't work.

The above seems to be the same as #11 .

@yisraeldov Should one have to manually set the var of ` org-sync-backend-alist`
for a hosted gitlab? I would think that should be set automatically.

> Labels would be nice, but I want to get the basic functionality merged in
> first.

Agreed.

> Weird I don't get that bug. Try using the version from git-hub master.

You mean with the github back end? I tried the default test repo on github and
had no such issue.

> What s the title when you go into your gitlab site ?

Not sure what you mean.

So created a new test project. Created 3 issues from the gitlab web site and did
an import into a new org file.

The import worked perfectly as per below:

```org
* Tasks
:PROPERTIES:
:url:      git.example.com/groupname/org-sync-test
:END:
** OPEN issue 1 created in gitlab
:PROPERTIES:
:web-url:  "https://git.example.com/groupname/org-sync-test/issues/3"
:id:       3
:date-modification: 2017-12-08T12:19:03+0200
:date-creation: 2017-12-08T12:19:03+0200
:END:
: description
** OPEN issue 2 created in gitlab
:PROPERTIES:
:web-url:  "https://git.example.com/groupname/org-sync-test/issues/2"
:id:       2
:date-modification: 2017-12-08T12:18:57+0200
:date-creation: 2017-12-08T12:18:57+0200
:END:
: description
** OPEN issue 1 created in gitlab
:PROPERTIES:
:web-url:  "https://git.example.com/groupname/org-sync-test/issues/1"
:id:       1
:date-modification: 2017-12-08T12:18:48+0200
:date-creation: 2017-12-08T12:18:48+0200
:END:
: description
```

After that, running a few org-synces, nothing changed, as expected.

Then created a new test issue 4 in web site. And ran a sync.

All worked just fine.

I then appended this to the local org mode file.
```org
** OPEN first issue created in emacs org mode
Hello world
```

And ran a `org-sync`. All worked just fine.

Strange I have no idea how I got that duplicated multiple `OPEN OPEN` thing last
time. Let's just forget about that now since I can't reproduce it.

I think this backend works very well. Though I'm seeing lots of bugs they mostly
seem related to `org-sync` in general.

Lastly, I wasn't able to close an issue.

I changed the `OPEN` in org mode to `CLOSED` and ran a sync. It ran through
quickly and reverted the local CLOSED back to OPEN and the remote gitlab
remained in an `OPEN` state.

But yeah, agree this looks good and hopefully this will be merged soon and
`org-sync` activity picked up as I there is a lot of potential for this being
pretty awesome.

@arbox hopefully you get a chance to look at this soon.

**Comment by yisraeldov** [comment]
Thank you for your comments. G-d willing,  I'll get back to you on Sunday.

**Comment by plajjan** [comment]
I too have tested this backend now and gotten it to work with my private gitlab instance. I can confirm that it's not possible to close issues. It's simply not implemented. I have added it to my version.

I also reworked how the (base-)URL parsing works so that it works better with a private gitlab instance or at least I think it does. I'm not sure if I should open a PR to @yisraeldov branch or if I should open yet another PR to this repo. My work is at https://github.com/plajjan/org-sync/tree/kll-gitlab

**Comment by plajjan** [comment]
deadline/due_date isn't working, I believe it's because it expects a complete datetimestamp but only gets a date

**Comment by yisraeldov** [comment]
Are you sure that closing issues doesn't work? I am pretty sure that I did
implement it.

**Comment by plajjan** [comment]
@yisraeldov yes, I'm sure. You need to set state_event=close in the PUT request to the gitlab API and that's just not there. I added it to my branch here: https://github.com/plajjan/org-sync/commit/ea6fe40b5654f2e9420f2583aded1f63fcf090ae

My branch is based on yours and it's up to date with what you've pushed publicly at least.

**Comment by plajjan** [comment]
@arbox how (in)active are you with this project these days? :)

It seems there are a few persons that are interested in getting GitLab support merged. I wonder should we try to get basic support merged to arbox/org-sync and then post any follow-up bug fixes later or should we try to converge and mature the gitlab patch set separately somewhere else before you would consider merging it here? I guess it mostly depends on how mature you want the patches to be before you merge and how much time you can/want to spend on reading the patches.

**Comment by arbox** [comment]
I think, we could do that if you've tested this stuff. I somebody is willing to co-maintain the project you can get commit rights.

**Comment by plajjan** [comment]
@arbox I would be willing to try and work a bit on this project to shape it up. I can't commit to being a long time maintainer. With my son born earlier this year I'm already low on time but I have a raging interest in using org-sync that I hope can compensate ;) I'm a newcomer to Emacs but I'm loving it. It's the environment I've been wanting for so long and org-mode is part of that. org-sync fits nicely since I work daily with GitLab so my motivation for fixing org-sync to work well with GitLab is quite high :)

**Comment by yisraeldov** [comment]
@plajjan That is strange, I remember testing it and it was working for closing tasks. I only tested on gitlab.com . 

I wrote the plugin for gitlab because I needed it working, but after no replies from the maintainer for a long time, I guess I just gave up. I have since moved on to other projects that aren't ( yet ) using gitlab for tasks. 

@plajjan If you want to make me a PR on my branch, I'll merge in your changes and update my PR #32

**Comment by arbox** [comment]
@plajjan nice :+1: Then I can add you as collaborator. 

As usual my beloved `sudo` citation:
```
We trust you have received the usual lecture from the local System
Administrator. It usually boils down to these three things:

    #1) Respect the privacy of others.
    #2) Think before you type.
    #3) With great power comes great responsibility.

root's password:
```

Thank you for your help!

**Comment by plajjan** [comment]
@arbox thanks! I'll try to behave :)

**Comment by yisraeldov** [comment]
@plajjan any chance we can get the PR merged in now ?

**Comment by yisraeldov** [comment]
@plajjan @arbox Any chance on getting the PR's merged in ?  Almost a month and no word.

**Comment by plajjan** [comment]
I've been traveling quite a bit so been a bit busy and had a lot to catch up on. I was hoping to spend some time on this the coming week :)

**Comment by plajjan** [comment]
Fixed through merge of #34 - closing.
